### ISSUE TITLE
as keyword

### ISSUE THREAD

**Issue by aliibrahim123** [tc39/proposal-type-annotations#150]
convert any type to a specific type.
example:
```
function example(a as Number) {
    return a+a
}

example('10')      //20
```

**Comment by ljharb** [comment]
That has runtime semantics, which nothing in this proposal does.

**Comment by lillallol** [comment]
> convert any type to a specific type.

This is already possible:

```js
function example(a) {
    if (typeof a === "string") return Number(a) + Number(a);
    if (typeof a === "number") return a+a;
    throw Error();//@TODO
}

example('10')      //20
```

so I see no reason for such a thing.

> That has runtime semantics, which nothing in this proposal does.

Correct if I am wrong, but that is not exactly what was said in the tc39 march meeting\[[1](https://github.com/tc39/notes/blob/main/meetings/2022-03/mar-29.md#types-as-comments-for-stage-1)\]\[[2](https://github.com/tc39/notes/blob/main/meetings/2022-03/mar-31.md#types-as-comments-continuation)\]. In fact being open to runtime semantics was a requirement to get stage1.

**Comment by zsakowitz** [comment]
This also has to TypeScript or Flow equivalent and was not part of the syntax presentation.

**Comment by orta** [comment]
> Correct if I am wrong

You're not wrong.

> RPR: The one on the slides and we will also consider runtime effects and simpler syntaxes.

I think one of the biggest arguments against _at least this syntax_ is that re-using `as` in a way which contradicts how it works in TypeScript is going to be semantic churn for anyone coming from that language fork (and probably a blocked on TS team agreeing to the proposal changes at a guess). 

You'd probably want to be looking at something like:

```ts
function example(a coerce Number) {
    return a+a
}

example('10')      //20
```

**Comment by lillallol** [comment]
> a blocker on TS team agreeing to the proposal changes

another condition for stage 1 was :

> WH: Okay. So are we open to simpler syntaxes? I hear that you've said yes, we're open to runtime semantics. Am I correct that we're also open to exploring very simple syntaxes that do not try to match what other existing forks do?
>
> DRR: I think that we're going to take those things into account, but yeah, I'm not against trying to understand what the use cases are and whether or not there are benefits and merits for one approach versus the other. ...

**Comment by orta** [comment]
I don't plan on debating more in this thread, but don't forget that this proposal needs buy in from the language forkers too - having a language feature which re-uses syntax which TypeScript has used for years and making it have unexpected side-effects like runtime changes are very unlikely to endorsed by the team. Which is why I offered an alternative way to look at the idea.

**Comment by lillallol** [comment]
> don't forget that this proposal needs buy in from the language forkers too 

I literally quoted a part from the tc39 march 2022 meeting,  in the previous comment, that shows that it is agreed that there is no such need.

> having a language feature which re-uses syntax which TypeScript has used for years and giving it side-effects like runtime changes are very unlikely to endorsed by the team

By which team? You mean TypeScript team? So what if the TypeScript team does not endorse it?
